src/pudl/transform/eiaaeo.py

```diff
@@ -0,0 +1,324 @@
+    return fields
+
+
+def filter_enrich_sanitize(
+    raw_df: pd.DataFrame, relevant_series_names: tuple[str]
+) -> pd.DataFrame:
+    """Basic cleaning steps common to all AEO tables.
+
+    1. Filter the AEO rows based on the series name
+    2. Break the series name and category names into useful fields
+    3. Sanitize strings & turn data values into a numeric field
+    4. Make some defensive checks about data from multiple sources that
+       *should* agree.
+    """
+    filtered = raw_df.loc[raw_df.series_name.str.startswith(relevant_series_names)]
+
+    enriched = pd.concat(
+        [
+            filtered,
+            get_series_info(filtered.series_name),
+            get_category_info(filtered.category_name),
+        ],
+        axis="columns",
+    ).drop(
+        columns=[
+            "series_name",
+            "category_name",
+            "variable_fields",
+        ]
+    )
+
+    sanitized = enriched.assign(
+        **{
+            colname: __sanitize_string(enriched[colname])
+            for colname in enriched.columns
+            if colname not in {"projection_year", "value"}
+        }
+    ).pipe(
+        lambda df: df.assign(
+            report_year_series=df.report_year_series.str.replace("aeo", ""),
+            value=pd.to_numeric(df.value.str.replace("- -", "")),
+        )
+    )
+
+    assert (
+        sanitized.region_series == sanitized.region_category
+    ).all(), "Series and category must agree on region!"
+    assert (
+        sanitized.case_series == sanitized.model_case_eiaaeo
+    ).all(), "Series and taxonomy must agree on case!"
+
+    # TODO 2024-04-20: one day we'll include report year in the extract step -
+    # at that point we'll have to compare that with the report year extracted
+    # from the series name.
+    return sanitized.drop(columns=["region_series", "case_series"]).rename(
+        columns={"region_category": "region", "report_year_series": "report_year"}
+    )
+
+
+def unstack(df: pd.DataFrame, eventual_pk: list[str]):
+    """Unstack the values by the various variable names provided.
+
+    2024-04-27: NB it's important that `projection_year` is the last level of
+    index, so that when we sort and diff() we are actually subtracting year
+    over year.
+    """
+    assert (
+        eventual_pk[-1] == "projection_year"
+    ), "Projection year must be last level of index!"
+    unstacked = (
+        df.set_index(eventual_pk + ["variable_name"])
+        .unstack(level="variable_name")
+        .sort_index()
+    )
+    unstacked.columns = unstacked.columns.droplevel()
+    unstacked.columns.name = None
+    return unstacked
+
+
+@asset(io_manager_key="pudl_io_manager")
+def core_eiaaeo__yearly_projected_generation_in_electric_sector_by_technology(
+    raw_eiaaeo__electric_power_projections_regional,
+):
+    """Projected net summer generation capacity and additions/retirements."""
+
+    def _drop_aggregated_generation_data(df: pd.DataFrame) -> pd.DataFrame:
```

Comment:
i could imagine some of these next few internal functions being generalized a bit and used for other assets given what we talked about in the schema process. but maybe not depending on the other tables! which is to say this looks cool and i'm sure you'll keep an eye towards generalizing when helpful while iterating over the other tables

Reply:
Yeah, I figured it would be good to encapsulate them within the function now and then move them out if necessary.